### ISSUE TITLE
OSO terminates Che workspaces 1h after their creation

### ISSUE THREAD

**Issue by slemeur** [openshiftio/openshift.io#3496]
Che workspace agent is crashing when used for this source code:
https://github.com/qodfathr/sentimentmeasurer

Workspace is unstable and becoming unresponsive. Displaying this error in the IDE:
![image pasted at 2018-5-7 13-24](https://user-images.githubusercontent.com/1636769/39715502-cad4b3ac-522d-11e8-9a62-c3bf783da2c8.png)

**Comment by ldimaggi** [comment]
I have seen this problem once today too - with a newly created Vert/X/REST quickstart - not sure of the sequence of actions - seems like it just happened after a period of inactivity in Che.

**Comment by eivantsov** [comment]
I think we are interested in java options for ws agent.

```
JAVA_OPTS=-XX:+UseG1GC -XX:+UseStringDeduplication -XX:MinHeapFreeRatio=20 -XX:MaxHeapFreeRatio=40 -XX:MaxRAM=1200m -Xms256m
```

```
CHE_WORKSPACE_WSAGENT__JAVA__OPTIONS=-XX:MaxRAM=600m -XX:MaxRAMFraction=1 -XX:+UseParallelGC -XX:MinHeapFreeRatio=10 -XX:MaxHeapFreeRatio=20 -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90 -Dsun.zip.disableMemoryMapping=true -Xms50m -Dfile.encoding=UTF8 -Djava.security.egd=file:/dev/./urandom
```

**Comment by qodfathr** [comment]
CHE_WORKSPACE_MAVEN__SERVER__JAVA__OPTIONS=-XX:MaxRAM=128m -XX:MaxRAMFraction=1 -XX:+UseParallelGC -XX:MinHeapFreeRatio=10 -XX:MaxHeapFreeRatio=20 -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90 -Dsun.zip.disableMemoryMapping=true -Xms20m -Djava
.security.egd=file:/dev/./urandom

**Comment by qodfathr** [comment]
![image](https://user-images.githubusercontent.com/16322190/39718253-97222686-51ea-11e8-85ac-c9728a7f7970.png)

**Comment by garagatyi** [comment]
So the reason might be that it takes a lot of CPU cycles for too long, WDYT?

**Comment by slemeur** [comment]
BTW .

when doing `env | grep JAVA`, we can see this:
`CHE_WORKSPACE_MAVEN__SERVER__JAVA__OPTIONS=-XX:MaxRAM=128m -XX:MaxRAMFraction=1 -XX:+UseParallelGC -XX:MinHeapFreeRatio=10 -XX:MaxHeapFreeRatio=20 -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90 -Dsun.zip.disableMemoryMapping=true -Xms20m -Djava .security.egd=file:/dev/./urandom `

But when going into the OSO Console we see this:
`CHE_WORKSPACE_WSAGENT__JAVA__OPTIONS=-XX:MaxRAM=600m -XX:MaxRAMFraction=1 -XX:+UseParallelGC -XX:MinHeapFreeRatio=10 -XX:MaxHeapFreeRatio=20 -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90 -Dsun.zip.disableMemoryMapping=true -Xms50m -Dfile.encoding=UTF8 -Djava.security.egd=file:/dev/./urandom`

**Comment by garagatyi** [comment]
`java -XX:+PrintGCDetails -XX:+Verbose   -XX:+UnlockExperimentalVMOptions -XX:+UseCGroupMemoryLimitForHeap -version ` should show memory settings

**Comment by slemeur** [comment]
we cant' run the command @garagatyi . Not available.

**Comment by garagatyi** [comment]
Didn't get what is not available. Can you elaborate?

**Comment by eivantsov** [comment]
```
[user@workspacei0l8wp6vp2s4r7pp projects]$ java -XX:+PrintGCDetails -XX:+Verbose -XX:+UnlockExperimentalVMOptions -XX:+UseCGroupMemoryLimitForHeap -version
Error: VM option 'Verbose' is develop and is available only in debug version of VM.
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
[user@workspacei0l8wp6vp2s4r7pp projects]$ java -XX:+PrintGCDetails -XX:+UnlockExperimentalVMOptions -XX:+UseCGroupMemoryLimitForHeap -version
openjdk version "1.8.0_161"
OpenJDK Runtime Environment (build 1.8.0_161-b14)
OpenJDK 64-Bit Server VM (build 25.161-b14, mixed mode)
Heap
 PSYoungGen      total 9728K, used 522K [0x00000000f5580000, 0x00000000f6000000, 0x0000000100000000)
  eden space 8704K, 6% used [0x00000000f5580000,0x00000000f5602aa0,0x00000000f5e00000)
  from space 1024K, 0% used [0x00000000f5f00000,0x00000000f5f00000,0x00000000f6000000)
  to   space 1024K, 0% used [0x00000000f5e00000,0x00000000f5e00000,0x00000000f5f00000)
 ParOldGen       total 22016K, used 0K [0x00000000e0000000, 0x00000000e1580000, 0x00000000f5580000)
  object space 22016K, 0% used [0x00000000e0000000,0x00000000e0000000,0x00000000e1580000)
 Metaspace       used 2425K, capacity 4480K, committed 4480K, reserved 1056768K
  class space    used 233K, capacity 384K, committed 384K, reserved 1048576K
```

**Comment by slemeur** [comment]
correct thanks @eivantsov

**Comment by slemeur** [comment]
This issue seems to be related to activeDeadline which is setup to `default` which means that the pod is going to be killed after one hour, even if there is the idlign mecanism in place in Che-master

**Comment by garagatyi** [comment]
I did an investigation of the issue and it seems that something weird is happening in communication between fabric8 k8s client and OSO cluster. We will need to investigate it more but for now, we would not be able to provide a solution in time. What I can suggest is setting the default value of activeDeadline on OSO clusters to something like 3 hours. Just for the demo purposes. 
@slemeur @kbsingh WDYT?

**Comment by riuvshin** [comment]
@garagatyi you are right, that is not a memory thing. this is related to settings of `activeDeadlineSeconds`. It is fixed for @qodfathr WS for demo.

**Comment by garagatyi** [comment]
An interesting thing I've discovered is that when Che creates WS pods in user-che namespace there is an `activeDeadline` property set, but when I create a pod in the same namespace from OSO cluster UI there is no `activeDeadline` and it seems that such pod can run forever. 
@riuvshin Do you know somebody from SD or OSO platform team who can help us figuring out whether it is a bug or not?

**Comment by l0rd** [comment]
An update on this issue. We figured out that Che workspaces are considered terminating pods whereas before they where considered non terminating:

![image](https://user-images.githubusercontent.com/606959/39827529-0b7a9346-536d-11e8-9565-c9dae9b76bb1.png)

The workaround for @qodfathr namespace has been to increase the default value for `activeDeadLine` (something that can be set at single namespace level).

We have been discussing about long term solution to avoid that the pods get killed (i.e. to avoid that OSO set `activeDeadline` for workspace pods) and possibilities are:
- change `Restart policy` from `Never` to `OnFailure`
- Use `deploymentConfig` to deploy workspace pods

**Comment by eivantsov** [comment]
@l0rd will restart policy really help? Even if a ws pod is restarted, it won't have any of the agents unless Che server re-bootstraps it again

**Comment by garagatyi** [comment]
@l0rd I think that workspace pods were always terminating, but, probably, default cluster-wide `activeDeadline` was introduced which caused the issue. 
I agree with @eivantsov that setting restart policy, for now, doesn't make sense because our workspaces can't restart, unfortunately. 
Deploying workspace with a deployment config seems very reasonable even without the need to mitigate the issue. I don't understand how usage of deployment config would help here since OS docs [say](https://docs.openshift.com/container-platform/3.3/admin_guide/managing_pods.html#manage-pods-limit-run-once-duration) that pod is terminating depending on its restart policy only. @l0rd can you elaborate?

**Comment by eivantsov** [comment]
As far as I understand, pods scheduled by deployments aren't subjected to such caps, otherwise Che server pod would have already had identical issues, while hasn't so far.

As to using deployments, not pods, for workspaces, it has its advantages of course, however, as I understand it, deployments are for apps that have their update and restart lifecycle. In case of a workspace pod, it is what it is - a short lived pod. For example, when a pipeline starts in OSIO, a new Jenkins slave is also a pod, not a deployment since it is going to be deleted once the job is done.

**Comment by aditya-konarde** [comment]
Is this still to be considered a SEV-1?

**Comment by mmclanerh** [comment]
can we get an update on where we are at with a fix here? is this a sev-1 item? this seems to fit more under a sev-3 description.

https://github.com/openshiftio/openshift.io/wiki/Labels

**Comment by riuvshin** [comment]
I think it was sev-1 / p0 just for because it was affecting demo.

**Comment by ibuziuk** [comment]
The issue is should be fixed on prod once we switch to 6.10.0 upstream che version (this releases contains fixes for regressions introduced by the switch from pods to deployments):
- https://github.com/eclipse/che/issues/10559
- https://github.com/eclipse/che/issues/10295

**Comment by ibuziuk** [comment]
closing. Fixes are now on prod
